File: boller-bot/src/commands/target.ts
import type { CommandHandler } from "@woife5/shared/lib/commands/types.d";
import { adminId } from "@woife5/shared/lib/utils/env.util";
import { resetTarget, setTarget } from "database/boller-target";
import { type ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder, type User } from "discord.js";
import { defaultEmbed } from "./embeds";

export const target: CommandHandler = {
    data: new SlashCommandBuilder()
        .setName("target")
        .setDescription("Set a target user. If no user is provided, the first user per guild will be joined after.")
        .addUserOption((option) => option.setName("user").setDescription("The user to set as target")),
    executeInteraction: async (interaction: ChatInputCommandInteraction): Promise<void> => {
        const user = interaction.options.getUser("user");

        if (interaction.user.id !== adminId) {
            await interaction.reply({
                embeds: [
                    new EmbedBuilder().setColor("Red").setDescription("Only the bot admin can set a target user."),
                ],
                ephemeral: true,
            });
            return;
        }

        await interaction.reply({ embeds: [await runCommand(user)] });
    },
};

async function runCommand(user: User | null) {
    if (!user) {
        try {
            await resetTarget();
            return defaultEmbed().setDescription(
                "Target has been reset. The first user per server will be joined after and bollered! :D After the last user leaves, the bot will also leave the channel.",
            );
        } catch (_) {
            return new EmbedBuilder().setColor("Red").setDescription("Failed to reset target. Please try again later.");
        }
    }

    try {
        await setTarget({ userId: user.id, userName: user.username });
        return defaultEmbed().setDescription(
            `Set the target user to <@${user.id}>. Will join each voice channel the user switches to and only leave after all users leave.`,
        );
    } catch (_) {
        return new EmbedBuilder().setColor("Red").setDescription("Failed to reset target. Please try again later.");
    }
}

File: boller-bot/commands/target.go
package commands

import (
	"context"
	"fmt"

	"github.com/bwmarrin/discordgo"
	"github.com/woife5/boller-bot/config"
	"github.com/woife5/boller-bot/database"
)

func (b *Bot) handleTarget(s *discordgo.Session, i *discordgo.InteractionCreate) {
	// Only the designated admin may use this command
	var userID string
	if i.Member != nil {
		userID = i.Member.User.ID
	} else if i.User != nil {
		userID = i.User.ID
	}

	if userID != b.WolfgangID {
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: "You don't have permission to use this command.",
				Flags:   discordgo.MessageFlagsEphemeral,
			},
		})
		return
	}

	opts := i.ApplicationCommandData().Options
	embed := defaultEmbed()
	embed.Thumbnail = &discordgo.MessageEmbedThumbnail{URL: config.BollerwagenLogoURL}

	// No user option supplied → reset
	if len(opts) == 0 || opts[0].Value == nil {
		if err := b.Targets.Reset(context.Background()); err != nil {
			embed.Description = fmt.Sprintf("Failed to reset target: %v", err)
		} else {
			embed.Description = "Target cleared. The bot will join any occupied voice channel."
		}
	} else {
		user := opts[0].UserValue(s)
		target := database.Target{
			UserID:   user.ID,
			UserName: user.Username,
		}
		if err := b.Targets.Set(context.Background(), target); err != nil {
			embed.Description = fmt.Sprintf("Failed to set target: %v", err)
		} else {
			embed.Description = fmt.Sprintf("Target set to <@%s> (%s).", user.ID, user.Username)
		}
	}

	s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseChannelMessageWithSource,
		Data: &discordgo.InteractionResponseData{
			Embeds: []*discordgo.MessageEmbed{embed},
		},
	})
}
